### PR TITLE
ISPN-14788 Allow to retrieve cache entry metadata version using the REST API

### PR DESCRIPTION
**Pull request by fax4ever**
https://issues.redhat.com/browse/ISPN-14788

### CODE REVIEW

File: server/rest/src/main/java/org/infinispan/rest/stream/CacheEntryStreamProcessor.java [hunk 1]
```diff
@@ -112,28 +116,51 @@ private static byte[] metadataBytes(CacheEntry<?, ?> currentEntry) {
...
+      Long version = null;
+      Integer topologyId = null;
+
+      if (entryVersion instanceof NumericVersion) {
+         version = ((NumericVersion) entryVersion).getVersion();
+      } else if (entryVersion instanceof SimpleClusteredVersion) {
+         version = ((SimpleClusteredVersion) entryVersion).getVersion();
+         topologyId = ((SimpleClusteredVersion) entryVersion).getTopologyId();
+      }
+      Json metadata = Json.object("timeToLiveSeconds", timeToLiveSeconds,
```
Comment (wburns) [inline server/rest/src/main/java/org/infinispan/rest/stream/CacheEntryStreamProcessor.java:150]: Using Json here seems like overkill.

Can't we just change it to use a StringBuilder and then do `sb.append(",\"version\":).append(version);` and then call `sb.toString().getBytes();` at the end?

File: server/rest/src/main/java/org/infinispan/rest/stream/CacheEntryStreamProcessor.java [hunk 2]
```diff
@@ -112,28 +116,51 @@ private static byte[] metadataBytes(CacheEntry<?, ?> currentEntry) {
       if (!(currentEntry instanceof InternalCacheEntry)) {
          return null;
       }
+
+      Metadata metadata = currentEntry.getMetadata();
```
Comment (pruivo) [inline server/rest/src/main/java/org/infinispan/rest/stream/CacheEntryStreamProcessor.java:120]: metadata can be null

File: server/rest/src/main/java/org/infinispan/rest/stream/CacheEntryStreamProcessor.java [hunk 3]
```diff
@@ -112,28 +116,51 @@ private static byte[] metadataBytes(CacheEntry<?, ?> currentEntry) {
...
       InternalCacheEntry ice = (InternalCacheEntry) currentEntry;
       // add metadata
-      long lifespanInSeconds = ice.getLifespan();
+      long lifespanInSeconds = metadata.lifespan();
```
Comment (pruivo) [inline server/rest/src/main/java/org/infinispan/rest/stream/CacheEntryStreamProcessor.java:123]: why change this line and the above?

File: server/rest/src/main/java/org/infinispan/rest/stream/CacheEntryStreamProcessor.java (hunk 1 above)
Comment (pruivo) [reply server/rest/src/main/java/org/infinispan/rest/stream/CacheEntryStreamProcessor.java:150]: agree

File: server/rest/src/test/java/org/infinispan/rest/resources/metedata/EntryVersionMetadataCollectionTest.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+import org.infinispan.metadata.EmbeddedMetadata;
+import org.infinispan.metadata.Metadata;
+import org.infinispan.rest.helper.RestServerHelper;
+import org.infinispan.test.SingleCacheManagerTest;
+import org.infinispan.test.fwk.TestCacheManagerFactory;
+import org.testng.annotations.Test;
+
+@Test(groups = "unit", testName = "rest.search.resources.metedata.EntryVersionMetadataCollectionTest")
+@TestForIssue(jiraKey = "ISPN-14788")
+public class EntryVersionMetadataCollectionTest extends SingleCacheManagerTest {
```
Comment (pruivo) [inline server/rest/src/test/java/org/infinispan/rest/resources/metedata/EntryVersionMetadataCollectionTest.java:37]: you don't need a new class. `org.infinispan.rest.resources.CacheResourceV2Test` tests all the methods related to the cache endpoint and the test can live there.

**Review by pruivo — CHANGES_REQUESTED** [review]

File: server/rest/src/main/java/org/infinispan/rest/stream/CacheEntryStreamProcessor.java (hunk 1 above)
Comment (fax4ever) [reply server/rest/src/main/java/org/infinispan/rest/stream/CacheEntryStreamProcessor.java:150]: In fact, I was in doubt if using a string builder and I agree with you both. I'm changing it. Thanks

File: server/rest/src/main/java/org/infinispan/rest/stream/CacheEntryStreamProcessor.java (hunk 3 above)
Comment (fax4ever) [reply server/rest/src/main/java/org/infinispan/rest/stream/CacheEntryStreamProcessor.java:123]: I thought it was better to use metadata instead. I'm reverting those changes.

File: server/rest/src/test/java/org/infinispan/rest/resources/metedata/EntryVersionMetadataCollectionTest.java (hunk 4 above)
Comment (fax4ever) [reply server/rest/src/test/java/org/infinispan/rest/resources/metedata/EntryVersionMetadataCollectionTest.java:37]: My idea was try to start to offload some code from `CacheResourceV2Test` class that has 1788 locs, but it is ok to add a method there.

**Comment by fax4ever** [comment]
@wburns and @pruivo thanks for the review, I applied the suggestions

**Review by wburns — APPROVED** [review]
LGTM

**Review by pruivo — APPROVED** [review]

**Comment by fax4ever** [comment]
Merged, thanks

**Pull request merged**